<a href="https://colab.research.google.com/github/RobertCurley/UCDPA_RobertCurley/blob/main/UCDPA_RobertCurley.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Abstract

Football is known for being statitical sport with a number of ratings for different aspects of the game. Football Manager 2023 is a computer game which allows you to manage a team. It has a rich database on the players and their ratings in different aspects of the game.

As I have an interest in the game of Football and gaming I have decided to look at these stats by players to delve deeper into the best player in the world according to Football Manager 2023.

Some of the findings which I can find from the dataset;

*   What country produces the players with the best finishing?
*   Do players with a higher heading rating tend to be taller?



#Introduction

*Explain why you chose this project use case*

# Data

Dataset being used in report;

*   *Football Manager 2023 Dataset* which can be viewed at:

https://www.kaggle.com/datasets/platinum22/foot-ball-manager-2023-dataset 

# Importing

Section covering the importing of the databases

In [2]:
# Import NumPy Package
import numpy as np

# Import Pandas Package
import pandas as pd

## Import *Football Manager 2023 Dataset*

In [3]:
# Upload CSV file of Football Manager 2023 from github repository

url = 'https://raw.githubusercontent.com/RobertCurley/UCDPA_RobertCurley/main/FM%202023.csv'

In [4]:
# View FM 2023.csv file in python

fm_2023_df = pd.read_csv(url) # Changed Name to FM_2023_df
fm_2023_df

Name  Position  Age   ca   pa     Nationality  \
0        Kevin De Bruyne  M/AM RLC   31  189  189         Belgium   
1          Kylian Mbappé   AM/S RL   23  188  197          France   
2     Robert Lewandowski         S   33  186  190          Poland   
3         Erling Haaland         S   22  185  195  Norway,England   
4          Mohamed Salah   AM/S RL   30  185  187           Egypt   
...                  ...       ...  ...  ...  ...             ...   
8447          Joe Ashton       D L   16   45  135         England   
8448          River Ries         S   17   45  135         Germany   
8449      Halilcan Doğan       D C   23   45  -75          Turkey   
8450        Adijat Sefer         S   17   45  135         Germany   
8451         Linus Urban         S   17   45  135         Germany   

                            Club  Corners  Crossing  Dribbling  ...  \
0                Manchester City       14        19         15  ...   
1            Paris Saint-Germain       13        13         18  ...   
2                      Barcelona        3         8         13  ...   
3                Manchester City        7        10         14  ...   
4                      Liverpool       12        14         17  ...   
...                          ...      ...       ...        ...  ...   
8447                     Burnley        3         4          5  ...   
8448               Karlsruher SC        2         2         11  ...   
8449  Osmaniyespor Futbol Kulübü        1         1          1  ...   
8450              TSG Hoffenheim        3         4         12  ...   
8451            SV Werder Bremen        4         2         10  ...   

      World reputation                           Race  RCA  Colour of skin  \
0                 9400              Northern_European  181               4   
1                 9248              African_Caribbean  172              13   
2                 9250              Northern_European  183               3   
3                 8750              Northern_European  185               2   
4                 8750  North_African__Middle_Eastern  181               9   
...                ...                            ...  ...             ...   
8447               236              Northern_European   41               3   
8448               462              Northern_European   33             255   
8449              1145              Northern_European   41               3   
8450               474              Northern_European   40             255   
8451               176              Northern_European   41             255   

      Date of birth  Number of national team appearances  \
0         1991/6/28                                   91   
1        1998/12/20                                   57   
2         1988/8/21                                  132   
3         2000/7/21                                   21   
4         1992/6/15                                   85   
...             ...                                  ...   
8447      2006/6/26                                    0   
8448       2005/4/3                                    0   
8449      1999/2/11                                    0   
8450      2005/1/20                                    0   
8451     2004/11/20                                    0   

      Goals scored for the national team     Salary  Rental club         UID  
0                                     24   394372.0          NaN    18004457  
1                                     27  1035616.0          NaN    85139014  
2                                     76   345204.0          NaN      719601  
3                                     20   394372.0          NaN    29179241  
4                                     47   405971.0          NaN    98028755  
...                                  ...        ...          ...         ...  
8447                                   0      203.0          NaN  2000210572  
8448                                   0      40

#Preparation

Section covering the preperation of the data.

*   *Create panda DataFrame*
*   *Sorting, Indexing, Grouping*
*   *Drop Duplicates, replace missing values*
*   *Merge Data Frames*

##Information on DataFrame

In [5]:
#Shows what the DataFrame looks like
fm_2023_df

Name  Position  Age   ca   pa     Nationality  \
0        Kevin De Bruyne  M/AM RLC   31  189  189         Belgium   
1          Kylian Mbappé   AM/S RL   23  188  197          France   
2     Robert Lewandowski         S   33  186  190          Poland   
3         Erling Haaland         S   22  185  195  Norway,England   
4          Mohamed Salah   AM/S RL   30  185  187           Egypt   
...                  ...       ...  ...  ...  ...             ...   
8447          Joe Ashton       D L   16   45  135         England   
8448          River Ries         S   17   45  135         Germany   
8449      Halilcan Doğan       D C   23   45  -75          Turkey   
8450        Adijat Sefer         S   17   45  135         Germany   
8451         Linus Urban         S   17   45  135         Germany   

                            Club  Corners  Crossing  Dribbling  ...  \
0                Manchester City       14        19         15  ...   
1            Paris Saint-Germain       13        13         18  ...   
2                      Barcelona        3         8         13  ...   
3                Manchester City        7        10         14  ...   
4                      Liverpool       12        14         17  ...   
...                          ...      ...       ...        ...  ...   
8447                     Burnley        3         4          5  ...   
8448               Karlsruher SC        2         2         11  ...   
8449  Osmaniyespor Futbol Kulübü        1         1          1  ...   
8450              TSG Hoffenheim        3         4         12  ...   
8451            SV Werder Bremen        4         2         10  ...   

      World reputation                           Race  RCA  Colour of skin  \
0                 9400              Northern_European  181               4   
1                 9248              African_Caribbean  172              13   
2                 9250              Northern_European  183               3   
3                 8750              Northern_European  185               2   
4                 8750  North_African__Middle_Eastern  181               9   
...                ...                            ...  ...             ...   
8447               236              Northern_European   41               3   
8448               462              Northern_European   33             255   
8449              1145              Northern_European   41               3   
8450               474              Northern_European   40             255   
8451               176              Northern_European   41             255   

      Date of birth  Number of national team appearances  \
0         1991/6/28                                   91   
1        1998/12/20                                   57   
2         1988/8/21                                  132   
3         2000/7/21                                   21   
4         1992/6/15                                   85   
...             ...                                  ...   
8447      2006/6/26                                    0   
8448       2005/4/3                                    0   
8449      1999/2/11                                    0   
8450      2005/1/20                                    0   
8451     2004/11/20                                    0   

      Goals scored for the national team     Salary  Rental club         UID  
0                                     24   394372.0          NaN    18004457  
1                                     27  1035616.0          NaN    85139014  
2                                     76   345204.0          NaN      719601  
3                                     20   394372.0          NaN    29179241  
4                                     47   405971.0          NaN    98028755  
...                                  ...        ...          ...         ...  
8447                                   0      203.0          NaN  2000210572  
8448                                   0      40

In [6]:
# Prints out full DataFrame
print(fm_2023_df)

                    Name  Position  Age   ca   pa     Nationality  \
0        Kevin De Bruyne  M/AM RLC   31  189  189         Belgium   
1          Kylian Mbappé   AM/S RL   23  188  197          France   
2     Robert Lewandowski         S   33  186  190          Poland   
3         Erling Haaland         S   22  185  195  Norway,England   
4          Mohamed Salah   AM/S RL   30  185  187           Egypt   
...                  ...       ...  ...  ...  ...             ...   
8447          Joe Ashton       D L   16   45  135         England   
8448          River Ries         S   17   45  135         Germany   
8449      Halilcan Doğan       D C   23   45  -75          Turkey   
8450        Adijat Sefer         S   17   45  135         Germany   
8451         Linus Urban         S   17   45  135         Germany   

                            Club  Corners  Crossing  Dribbling  ...  \
0                Manchester City       14        19         15  ...   
1            Paris Saint-Germ

In [7]:
# Returns the shape of the dataframe
fm_2023_df.shape 

(8452, 98)

In [8]:
# Returns the information present in the dataframe
fm_2023_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8452 entries, 0 to 8451
Data columns (total 98 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Name                                 8452 non-null   object 
 1   Position                             8452 non-null   object 
 2   Age                                  8452 non-null   int64  
 3   ca                                   8452 non-null   int64  
 4   pa                                   8452 non-null   int64  
 5   Nationality                          8452 non-null   object 
 6   Club                                 8345 non-null   object 
 7   Corners                              8452 non-null   int64  
 8   Crossing                             8452 non-null   int64  
 9   Dribbling                            8452 non-null   int64  
 10  Finishing                            8452 non-null   int64  
 11  First Touch                   

In [9]:
# Returns List of Title of all Columns in DataFrame
fm_2023_df.columns

Index(['Name', 'Position', 'Age', 'ca', 'pa', 'Nationality', 'Club', 'Corners',
       'Crossing', 'Dribbling', 'Finishing', 'First Touch', 'Free Kick Taking',
       'Heading', 'Long Shots', 'Long Throws', 'Marking', 'Passing',
       'Penalty Taking', 'Tackling', 'Technique', 'Aggressiion',
       'Anticipation', 'Bravery', 'Composure', 'Concentration', 'Vision',
       'Decision', 'Determination', 'Flair', 'Leadership', 'Off The Ball',
       'Position.1', 'Teamwork', 'Work Rate', 'Acceleration', 'Agility',
       'Balance', 'Jumping Reach', 'Natural Fitness', 'Pace', 'Stamina',
       'Strength', 'Stability', 'Foul', 'Contest performance', 'Injury',
       'diversity', 'Aerial Reach', 'Command Of Area', 'Communication',
       'Eccentricity', 'Handling', 'Kicking', 'One On Ones', 'Reflexes',
       'Rushing Out', 'Punching', 'Throwing', 'Adaptation', 'Ambition',
       'Argue', 'Loyal', 'Resistant to stress', 'Professional',
       'Sportsmanship', 'Emotional control', 'GK', 'DL', 

##Creating DataFrames

Created DataFrames to view the dataset in an easier to read way.

In [10]:
# Slice Data in DataFrame (fm_2023_df) and name KeyDetails_fm_2023_df
# Key Details : Name, Position, Nationality, Club, Age, Height, Weight, salary ca & pa

KeyDetails_fm_2023_df = fm_2023_df.loc[:,['Name', 'Position', 'Nationality', 'Club', 
                                          'Age', 'Height', 'Weight', 'Salary', 'ca', 'pa']].set_index('Name')

print(KeyDetails_fm_2023_df)

                    Position     Nationality                        Club  Age  \
Name                                                                            
Kevin De Bruyne     M/AM RLC         Belgium             Manchester City   31   
Kylian Mbappé        AM/S RL          France         Paris Saint-Germain   23   
Robert Lewandowski         S          Poland                   Barcelona   33   
Erling Haaland             S  Norway,England             Manchester City   22   
Mohamed Salah        AM/S RL           Egypt                   Liverpool   30   
...                      ...             ...                         ...  ...   
Joe Ashton               D L         England                     Burnley   16   
River Ries                 S         Germany               Karlsruher SC   17   
Halilcan Doğan           D C          Turkey  Osmaniyespor Futbol Kulübü   23   
Adijat Sefer               S         Germany              TSG Hoffenheim   17   
Linus Urban                S

In [11]:
# Slice Data in DataFrame (fm_2023_df) based on key stats for scoring and name it scoring_fm_2023_df
# Key Stats : Name, Dribbling, Finishing, First Touch, Free Kick Taking, Heading, Long Shots, Penalty Taking

scoring_fm_2023_df = fm_2023_df.loc[:, ['Name', 'Dribbling', 'Finishing', 'First Touch', 'Free Kick Taking', 'Heading',
                                        'Long Shots', 'Penalty Taking']].set_index('Name')
print(scoring_fm_2023_df)

                    Dribbling  Finishing  First Touch  Free Kick Taking  \
Name                                                                      
Kevin De Bruyne            15         16           16                17   
Kylian Mbappé              18         17           18                12   
Robert Lewandowski         13         19           18                15   
Erling Haaland             14         18           16                13   
Mohamed Salah              17         17           17                12   
...                       ...        ...          ...               ...   
Joe Ashton                  5          4            4                 1   
River Ries                 11         11           11                 2   
Halilcan Doğan              1          2            4                 1   
Adijat Sefer               12          7            8                 3   
Linus Urban                10         13            8                 4   

                    Head

##Merge DataFrames

Merge the 2 DataFrames Created

In [12]:
# Create Summary List based of 2 created DataFrames (KeyDetails_fm_2023_df & scoring_fm_2023_df) and name Summary_fm_2023_df

Summary_fm_2023_df = KeyDetails_fm_2023_df.merge(scoring_fm_2023_df, on = 'Name')
print(Summary_fm_2023_df)

                    Position     Nationality                        Club  Age  \
Name                                                                            
Kevin De Bruyne     M/AM RLC         Belgium             Manchester City   31   
Kylian Mbappé        AM/S RL          France         Paris Saint-Germain   23   
Robert Lewandowski         S          Poland                   Barcelona   33   
Erling Haaland             S  Norway,England             Manchester City   22   
Mohamed Salah        AM/S RL           Egypt                   Liverpool   30   
...                      ...             ...                         ...  ...   
Joe Ashton               D L         England                     Burnley   16   
River Ries                 S         Germany               Karlsruher SC   17   
Halilcan Doğan           D C          Turkey  Osmaniyespor Futbol Kulübü   23   
Adijat Sefer               S         Germany              TSG Hoffenheim   17   
Linus Urban                S

##Sorting DataFrames



In [13]:
# Sort Summary_fm_2023_df by Finishing

Summary_fm_2023_df.sort_values(by='Finishing', ascending=False).iloc[:6,10:11]

Finishing
Name                         
Zlatan Ibrahimović         19
Robert Lewandowski         19
Harry Kane                 19
Cristiano Ronaldo          19
Erling Haaland             18
Heung-Min Son              18

In [14]:
# Sort Summary_fm_2023_df by pa

Summary_fm_2023_df.sort_values(by="pa", ascending=False).iloc[:10, 6:9]

Salary   ca   pa
Name                                   
Lionel Messi         776712.0  180  200
Kylian Mbappé       1035616.0  188  197
Cristiano Ronaldo    556760.0  158  196
Erling Haaland       394372.0  185  195
Gianluigi Buffon      30239.0  125  194
Manuel Neuer         345204.0  175  193
Robert Lewandowski   345204.0  186  190
Neymar               939648.0  177  190
Kevin De Bruyne      394372.0  189  189
Luis Suárez            8121.0  147  188

##Grouping DataFrames

In [15]:
# Group Summary_fm_2023_df by Club

Club_df = Summary_fm_2023_df.groupby(by="Club").mean()
Club_df.sort_values(by="Age", ascending=False).iloc[:10,0:5]

Age  Height  Weight   Salary     ca
Club                                                                
Hughes Foot Ball Club           39.0   178.0    73.0  13764.0  124.0
FIG                             38.0   185.0    87.0   4858.0  120.0
FC Bayern Alzenau               38.0   185.0    81.0   9697.0  120.0
Deportes Magallanes             38.0   183.0    75.0   9475.0  120.0
Castellón                       37.0   173.0    64.0   6891.0  127.0
LKS Goczałkowice-Zdrój          37.0   184.0    79.0     28.0  125.0
Sporting FC (CRC)               37.0   183.0    82.0   9615.0  120.0
CSD Municipal                   36.0   193.0    82.0    849.0  122.0
Grazer Athletiksport Klub 1902  36.0   176.0    74.0  10323.0  121.0
Tauro FC                        36.0   187.0    80.0    931.0  121.0

In [16]:
# Group Summary_fm_2023_df by Age

Age_df = Summary_fm_2023_df.groupby(by="Age").mean()
Age_df.sort_values(by="Age", ascending=True).iloc[:10,2:3]

Salary
Age              
15     102.000000
16     299.800000
17    1926.606383
18    3014.236607
19    6427.360262
20    7927.312757
21   10599.637037
22   17299.951049
23   20894.114458
24   25934.813483

##Dropping Duplicates

* Drop duplicates within 'Club' column in fm_2023_df (example purpose only)
* Drop Null values within 'Salary' column in Summary_fm_2023_df

###Dropping Duplicates within 'Club' column in Summary_fm_2023_df (Example purpose only)

In [17]:
#Dropping Duplicates within the 'Club' Column in fm_2023_df

Drop_duplicates_df = fm_2023_df.drop_duplicates(subset="Club").iloc[:,6:7]
print(Drop_duplicates_df)
Drop_duplicates_df.shape

                            Club
0                Manchester City
1            Paris Saint-Germain
2                      Barcelona
4                      Liverpool
5                      R. Madrid
...                          ...
8407             KF Ramiz Sadiku
8410         Dalian Professional
8416            Al-Muharraq Club
8421           Chengdu Rongcheng
8449  Osmaniyespor Futbol Kulübü

[919 rows x 1 columns]


(919, 1)

In [18]:
#Check if the number of values in Drop_duplicates_df matches the number of Unique values in 'Club' column from fm_2023_df

Drop_duplicates_df["Club"].count() == fm_2023_df["Club"].nunique()

True

###Drop Null values within Summary_fm_2023_df

In [19]:
#What columns contain Null values in Summary_fm_2023_df

Summary_fm_2023_df.isnull().sum()

Position              0
Nationality           0
Club                107
Age                   0
Height                0
Weight                0
Salary              107
ca                    0
pa                    0
Dribbling             0
Finishing             0
First Touch           0
Free Kick Taking      0
Heading               0
Long Shots            0
Penalty Taking        0
dtype: int64

In [20]:
#Show how many missing values are within 'Salary' column in fm_2023_df

print("Salary column contains: ", Summary_fm_2023_df["Salary"].isnull().sum(), "missing values")
print("Age column contains: ", Summary_fm_2023_df["Age"].isnull().sum(), "missing values")

Salary column contains:  107 missing values
Age column contains:  0 missing values


In [21]:
#Drop Null values within Summary_fm_2023_df

Summary_fm_2023_df = Summary_fm_2023_df.dropna()
Summary_fm_2023_df.isnull().sum()

Position            0
Nationality         0
Club                0
Age                 0
Height              0
Weight              0
Salary              0
ca                  0
pa                  0
Dribbling           0
Finishing           0
First Touch         0
Free Kick Taking    0
Heading             0
Long Shots          0
Penalty Taking      0
dtype: int64

##Replacing missing values (Used for example purpose\)

Replace Null Values within 'Salary' column with the mean value based on 'Age'




In [22]:
#Create DataFrame showing the mean values for 'Salary' Column by 'Age'

Salary_mean_df = Age_df.sort_values(by="Age", ascending=True).iloc[:,2:3] #Age_df was created in Grouping DataFrames section
print(Salary_mean_df)

           Salary
Age              
15     102.000000
16     299.800000
17    1926.606383
18    3014.236607
19    6427.360262
20    7927.312757
21   10599.637037
22   17299.951049
23   20894.114458
24   25934.813483
25   28423.075697
26   29686.011364
27   28680.973577
28   30851.402490
29   30421.368644
30   35533.753181
31   34108.674863
32   27846.211009
33   26717.555985
34   27759.793103
35   29568.256410
36   37036.671053
37   32844.717391
38   16980.000000
39   18708.533333
40   20265.285714
41   12002.000000
42   12930.500000
44   30239.000000


In [23]:
#Convert Salary_mean_df to dictionary

Salary_mean_dict = Salary_mean_df.to_dict()

In [24]:
#Fill in null values in 'Salary' column in fm_2023_df using Salary_mean_df dict

print("The number of null values within the 'Salary' column was :", fm_2023_df["Salary"].isnull().sum())
Replaced_Salary = fm_2023_df.set_index("Age").fillna(value=Salary_mean_dict)
print("The number of null values within the 'Salary' column after formula is :", Replaced_Salary["Salary"].isnull().sum())

The number of null values within the 'Salary' column was : 107
The number of null values within the 'Salary' column after formula is : 0


#Analysis

* Conditional Statements, looping, groupby
* Define a custom function to create reusable code
* Use NumPy functions
* Dictionary or Lists

##Conditional Statement

In [25]:
#Recall Summary_fm_2023_df

print(Summary_fm_2023_df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 8345 entries, Kevin De Bruyne to Linus Urban
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Position          8345 non-null   object 
 1   Nationality       8345 non-null   object 
 2   Club              8345 non-null   object 
 3   Age               8345 non-null   int64  
 4   Height            8345 non-null   int64  
 5   Weight            8345 non-null   int64  
 6   Salary            8345 non-null   float64
 7   ca                8345 non-null   int64  
 8   pa                8345 non-null   int64  
 9   Dribbling         8345 non-null   int64  
 10  Finishing         8345 non-null   int64  
 11  First Touch       8345 non-null   int64  
 12  Free Kick Taking  8345 non-null   int64  
 13  Heading           8345 non-null   int64  
 14  Long Shots        8345 non-null   int64  
 15  Penalty Taking    8345 non-null   int64  
dtypes: float64(1), int64(12), 

In [26]:
#Create conditional statement to return the amount of players with Salary greater than 50000

Summary_fm_2023_df.loc[Summary_fm_2023_df['Salary'] >= 50000, 'Salary >= 50,000'] = 'True'

print("The total amount of players in the database is :", Summary_fm_2023_df["Club"].count())
print("The amount of players with a salary over 50,000 is :", Summary_fm_2023_df['Salary >= 50,000'].notnull().sum())

Percentage_Salary_Insight = round(((Summary_fm_2023_df['Salary >= 50,000'].notnull().sum() / Summary_fm_2023_df["Club"].count()) * 100), 1)

print("The percentage of players in football manager 2023 that earn greater than or equal to 50,000 is :", Percentage_Salary_Insight,"%")

The total amount of players in the database is : 8345
The amount of players with a salary over 50,000 is : 814
The percentage of players in football manager 2023 that earn greater than or equal to 50,000 is : 9.8 %


###Looping

In [27]:
#Code

#Visualisation

* Generate at least two charts using Matplotlib or seaborn
* https://pandas.pydata.org/pandas-docs/version/0.13/visualization.html 

##Create Scatter Plot looking at Height vs Heading



In [35]:
#Import MatPlotLib
import matplotlib.pyplot as plt

Height = Summary_fm_2023_df["Height"].reset_index(drop=True)
Heading = Summary_fm_2023_df["Heading"]
print(Height)
print(Heading)


0       181
1       178
2       185
3       195
4       175
       ... 
8340    181
8341    184
8342    190
8343    184
8344    174
Name: Height, Length: 8345, dtype: int64
Name
Kevin De Bruyne        6
Kylian Mbappé          7
Robert Lewandowski    17
Erling Haaland        15
Mohamed Salah         10
                      ..
Joe Ashton             6
River Ries             6
Halilcan Doğan        12
Adijat Sefer          12
Linus Urban           13
Name: Heading, Length: 8345, dtype: int64


#Insights

* Derive five valuable insights from the analysis
* Justify your insights with reference to the charts or analysis

##Valuable insights from the analysis

1.  Only 9.8% of players in Football Manager 2023 earn greater than or equal to 50,000 (814 out of a total 8345) 
2.   
3. 
4. 
5. 

In [29]:
#Code

#Machine Learning

* Describe what kind of prediction you could perform in future using machine learning and/or deep learning.
* Would you use classification or regression methods

##Prediction model